### PR TITLE
blog: Streamlining access to Grafana & Prometheus

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
This is part 1 of a 2-part blog post series on a recent set of changes to improve Mattermost Performance Monitoring. This particular blog posts focuses on the developer tooling improvements that make performance monitoring more accessible.

I considered trying to write about both the new dynamic annotations /and/ the developer changes in one go, but the topics felt forced being together. I'm not opposed to just cutting this half of the content altogether if we don't think it makes a good blog post, either -- let me know your thoughts!

#### Ticket Link
None.

### CODE REVIEW

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (jespino) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:47]: One question, do you have to enable "Metrics" in your mattermost to make it work, right? Probably adding that info here would be great, something like "Set your `MetricsSettings.Enable` to `true` in your configuration file (or through the admin console)".

Comment (lieut-data) [reply site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:47]: Excellent point! I'll also add a note here about having to restart the server when installing a license -- for some reason, metrics (even if enabled) don't start until the next server start.

Comment (streamer45) [reply site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:47]: From my experience it's sufficient to trigger a config save to make the endpoint work. This is assuming the server build is enterprise ready. I think we are missing some logic to start the server when a valid license is added. It might be worth consider fixing this edge case.

Comment (lieut-data) [reply site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:47]: Indeed! I'll look at that edge cases vs. permanently documenting the pain point here. Otherwise, I've pushed changes to document enabling the metrics. Thanks!

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (streamer45) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:22]:
```suggestion
    - `minio` to expose an S3-compatible interface to your local disk for hosting uploaded files
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (streamer45) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:32]: I am obviously biased but since metrics are also a requirement for running certain type of load-tests (making full use of the coordinator) it could be nice to mention that here as an added benefit of having this.

**Review by streamer45 — APPROVED** [review]
Looks great :+1: Left a couple of non-blocking suggestions.

**Review by jespino — APPROVED** [review]
LGTM

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:17]:
```suggestion
* Downloading a suitable version of `mmctl` for API-driven command line interaction.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:18]:
```suggestion
* Checking your installed Go version to avoid cryptic error messages.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:19]:
```suggestion
* Downloading and starting various Docker containers:
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:20]:
```suggestion
    - `mysql` and `postgres`, both for a backend to the server and also to automate tests.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:21]:
```suggestion
    - `inbucket` to simplify email testing without routing outside your machine.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md]:
```suggestion
    - `minio` to expose an S3-compatible interface to your local disk for hosting uploaded files.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:23]:
```suggestion
    - `openldap` to simplify testing alternate modes of authentication (requires access to the enterprise source).
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:24]:
```suggestion
    - `elasticsearch` to expose an improved search backend for posts and channels (requires access to the Enterprise source code).
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:25]:
```suggestion
* Linking your server to an automatically-detected `mattermost-webapp` directory.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:26]:
```suggestion
* Optionally starting the server in the foreground if `RUN_SERVER_IN_BACKGROUND` is `false`.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:27]:
```suggestion
* Piping the structured server logs through a `logrus` decorator for easier reading.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:28]:
```suggestion
* Invoking the `go` compiler to run the server code in your local repository.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:30]:
```suggestion
### Adding Prometheus and Grafana
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:34]:
```suggestion
Since the corresponding Mattermost functionality is only available with access to the Enterprise source code, these containers aren't enabled by default. To turn them on, export the following environment variable into your shell:
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:38]:
```suggestion
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:39]:
```suggestion
Alternatively, you can follow the instructions in [config.mk](https://github.com/mattermost/mattermost-server/blob/master/config.mk). Note that it's not necessary to specify `minio`, `openldap`, or `elasticsearch` - these are added automatically when an Enterprise-enabled repository is detected.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:41]:
```suggestion
The build tooling does more than just spin up two new containers, however. [Various configuration files](https://github.com/mattermost/mattermost-server/tree/master/build/docker) also automate the following:

```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:42]:
```suggestion
* Configuring Prometheus to scrape the Mattermost `:8067/metrics` endpoint (requires access to the Enterprise source code).
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:43]:
```suggestion
* Enabling anonymous access to Grafana (you can still log in with the default `admin`/`admin` credentials if needed).
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:44]:
```suggestion
* Configuring Grafana with the containerized Prometheus as the default data source.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:45]:
```suggestion
* Installing our canonical Grafana dashboards ([Mattermost Performance KPI Metrics](https://grafana.com/grafana/dashboards/2539), [Mattermost Performance Monitoring](https://grafana.com/grafana/dashboards/2542), [Mattermost Performance Monitoring (Bonus Metrics)](https://grafana.com/grafana/dashboards/2545)).
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:57]:
```suggestion
Although the tooling is all set up to scrape metrics and display the dashboards, you'll need to do some one-time work to enable metrics within the Mattermost server.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:59]:
```suggestion
First, be sure you have an Enterprise license installed. Staff members should have access to the shared developer license, but you can also request a trial license in-product. Browse to `/admin_console/about/license` on your local Mattermost instance to set up the license.
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:67]:
```suggestion
These changes weren't just made in isolation - in fact, this is all just infrastructure work for another recent project to improve Mattermost performance monitoring. Stay tuned for a follow-up blog post!
```

File: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md
Comment (justinegeffen) [inline site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md:11]:
```suggestion
summary: With access to the Enterprise source code, the developer build tooling now automates the setup of Prometheus and Grafana for performance monitoring. Even the canonical Grafana dashboards are setup without any manual configuration required!
```

**Review by justinegeffen — COMMENTED** [review]
This is really well-written and clearly explained. I've made some suggestions around general formatting and punctuation but no major content changes. Happy to discuss further, if you have any questions. :)

**Comment by lieut-data** [comment]
Brilliant! Thanks, @justinegeffen :)

**Comment by lieut-data** [comment]
@JustinReynolds-MM, I think we're close on wrapping up the effort for this blog post. Let me know if I need to do anything other than merge to the developer-only blog.

**Review by justinegeffen — APPROVED** [review]

**Comment by JustinReynolds-MM** [comment]
@lieut-data I'll load this into WordPress, and get it queued up for publish on Tuesday 2/2. I'll let you know once we're live on mattermost.com, then you can merge to the developer blog. Stay tuned !

**Comment by JustinReynolds-MM** [comment]
@lieut-data this one is now live: https://mattermost.com/blog/streamlining-developer-access-to-prometheus-and-grafana/

**Comment by lieut-data** [comment]
Thanks, @JustinReynolds-MM! The org reset seems to have broken the previously approved reviews -- can I get a quick approval from folks again? (@justinegeffen / @JustinReynolds-MM might suffice?)

**Review by justinegeffen — APPROVED** [review]

**Comment by justinegeffen** [comment]
@lieut-data, looks like @jaydeland needs to approve as code owner.

**Review by jaydeland — APPROVED** [review]

**Pull request merged**
